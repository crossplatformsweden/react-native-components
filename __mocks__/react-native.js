module.exports = ({
  StyleSheet: {
    create: (styles) => styles,
  },
  Platform: {
    OS: 'ios',
  },
  Dimensions: {
    get: () => ({
      width: 1,
      height: 1,
    }),
  },
  View: 'View',
  Text: 'Text',
  Button: 'Button',
  ScrollView: 'View',
  Linking: { openURL: (url) => {console.log('Lingking.openURL', url)}},
  AsyncStorage: {
    setItem: () => new Promise((resolve) => resolve(), null),
    getItem: (key) =>
      new Promise((resolve) => resolve(`{"key": "${key}"}`), null),
    removeItem: (key, callback) => {
      if (callback) {
        callback();
      }
      return new Promise((resolve) => resolve(), null);
    },
  },
  Animated: {
    View: 'View',
    Value: jest.fn()
  },
  PanResponder: {
    create: () => 'View'
  },
  PanResponderInstance: 'View',
  TouchableOpacity: 'View',
  FlatList: 'View',
});
