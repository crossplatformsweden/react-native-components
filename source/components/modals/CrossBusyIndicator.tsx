import React from 'react';
import {
  View,
  Text,
  TouchableOpacity,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';
import { Colors } from '../../styles';
import Modal from 'react-native-modal';
import {
  CrossSpinnerType,
  CrossSpinner,
  ICrossSpinnerProps,
} from '../animations/CrossSpinner';
import styles from '../../styles';

export type AnimationConfig = string | { from: Object; to: Object };

export type Orientation =
  | 'portrait'
  | 'portrait-upside-down'
  | 'landscape'
  | 'landscape-left'
  | 'landscape-right';

/**
 * Representation of react-native-modal props. All props except `isVisible` and `children` from:
 *
 * https://github.com/react-native-community/react-native-modal#available-props
 */
export interface IModalProps {
  animationIn?: AnimationConfig;
  animationInTiming?: number;
  animationOut?: AnimationConfig;
  animationOutTiming?: number;
  avoidKeyboard?: boolean;
  backdropColor?: string;
  backdropOpacity?: number;
  backdropTransitionInTiming?: number;
  backdropTransitionOutTiming?: number;
  useNativeDriver?: boolean;
  deviceHeight?: number;
  deviceWidth?: number;
  hideModalContentWhileAnimating?: boolean;
  onModalShow?: () => void;
  onModalHide?: () => void;
  onBackButtonPress?: () => void;
  onBackdropPress?: () => void;
  onSwipe?: () => void;
  swipeThreshold?: number;
  style?: StyleProp<ViewStyle>;
  swipeDirection?: 'up' | 'down' | 'left' | 'right';
  scrollTo?: (e: any) => void;
  scrollOffset?: number;
  scrollOffsetMax?: number;
  supportedOrientations?: Orientation[];
  onDismiss?: () => void;
  onShow?: () => void;
  hardwareAccelerated?: boolean;
  onOrientationChange?: (orientation: 'portrait' | 'landscape') => void;
  presentationStyle?:
    | 'fullScreen'
    | 'pageSheet'
    | 'formSheet'
    | 'overFullScreen';
}

/**
 * Describes the props for {@link CrossBusyIndicator}
 *
 * @export
 * @interface IBusyIndicatorProps
 */
export interface IBusyIndicatorProps {
  /**
   * Determines if the indicator is visible
   *
   * @type {boolean}
   * @memberof IBusyIndicatorProps
   */
  isBusy: boolean;
  /**
   * Optional message to show while busy. Defaults to empty string.
   *
   * @type {string}
   * @memberof IBusyIndicatorProps
   */
  message?: string | undefined;
  /**
   * Type of spinner to use
   *
   * @type {CrossSpinnerType}
   * @memberof IBusyIndicatorProps
   */
  type?: CrossSpinnerType | undefined;
  /**
   * Optional test id to assign this component. Default is "1"
   *
   * @type {string}
   * @memberof IBusyIndicatorProps
   */
  testID?: string | undefined;
  /**
   * Determines if the modal is cancellable. Otherwise visibility is controlled by {@link isBusy}
   */
  isCancelButtonVisible: boolean;
  /**
   * Optional callback for user cancel event.
   */
  onCancel?: () => void;
  /**
   * Text for cancel button. Default is "Cancel"
   */
  cancelText?: string | undefined;
  /**
   * Optional custom style for the {@link cancelText}
   */
  cancelTextStyle?: StyleProp<TextStyle> | undefined;
  /**
   * Optional custom style for the modal view
   */
  modalStyle?: StyleProp<ViewStyle> | undefined;
  /**
   * Optional custom style for the inner view, contained in the modal
   */
  innerViewStyle?: StyleProp<ViewStyle> | undefined;
  /**
   * Optional custom style for the {@link message}
   */
  messageStyle?: StyleProp<TextStyle> | undefined;
  /**
   * Optional properties to configure the spinner
   *
   * See {@link ICrossSpinnerProps}
   */
  spinnerProps?: ICrossSpinnerProps | undefined;
  /**
   * Optional properties to customize the modal. All props except `isVisible` and `children` from
   *
   * https://github.com/react-native-community/react-native-modal#available-props
   */
  modalProps?: IModalProps | undefined;
}

/**
 * @description
 * Displays a modal with busy indicator (spinner) and message when {@link IBusyIndicatorProps.isBusy} is `true`.
 *
 * Properties are {@link IBusyIndicatorProps}.
 *
 * Customize the modal through {@link IBusyIndicatorProps.modalProps}.
 *
 * Also supports cancelling the modal using properties
 * {@link IBusyIndicatorProps.isCancelButtonVisible},
 * {@link IBusyIndicatorProps.cancelText} and
 * {@link IBusyIndicatorProps.onCancel}.
 * This does however not hide the modal.
 *
 * @example <caption>Modal with cancel</caption>
 *  <CrossBusyIndicator
 *         isBusy={this.state.isBusy}
 *         type='PacmanIndicator'
 *         isCancelButtonVisible={true}
 *         message="Loading.."
 *         onCancel={() => this.setState({ isBusy: false })}
 *       />
 *
 * @example <caption>Non cancellable and custom styles</caption>
 *       <CrossBusyIndicator
 *          spinnerProps={{
 *            color: 'red',
 *            type: 'WaveIndicator'
 *          }}
 *          messageStyle={{color: 'red'}}
 *          isBusy={this.state.isBusy2}
 *          isCancelButtonVisible={false}
 *          message="Resistance is futile"
 *        />
 *
 * @example <caption>Customize modal props</caption>
 *       <CrossBusyIndicator
 *          modalProps={{
 *            swipeDirection: 'up',
 *            backdropColor: 'blue'
 *          }}
 *          isBusy={this.state.isBusy2}
 *          isCancelButtonVisible={false}
 *          message="Busy busy busy.."
 *        />
 */
export class CrossBusyIndicator extends React.Component<IBusyIndicatorProps> {
  constructor(props: IBusyIndicatorProps) {
    super(props);
    this.onCancel = this.onCancel.bind(this);
  }

  onCancel() {
    if (!this.props.isCancelButtonVisible) {
      return;
    }

    if (this.props.onCancel) {
      this.props.onCancel();
    }
  }

  render() {
    return (
      <View testID={this.props.testID || '1'}>
        <Modal
          {...this.props.modalProps}
          style={
            this.props.modalStyle
              ? [styles.absoluteCentered, this.props.modalStyle]
              : styles.absoluteCentered
          }
          isVisible={this.props.isBusy}
          onDismiss={this.onCancel}
        >
          <View
            style={
              this.props.innerViewStyle
                ? [
                    styles.container,
                    styles.paddingDefault,
                    this.props.innerViewStyle,
                  ]
                : [styles.container, styles.paddingDefault]
            }
          >
            <View style={[styles.columnContentTopCenter]}>
              <CrossSpinner
                type={this.props.type || 'MaterialIndicator'}
                style={styles.spinner}
                {...this.props.spinnerProps}
              />
            </View>

            <View style={[styles.columnContentTopCenter]}>
              <Text style={[styles.textSpinner, this.props.messageStyle]}>
                {this.props.message || ''}
              </Text>
              {this.props.isCancelButtonVisible ? (
                <TouchableOpacity onPress={this.onCancel}>
                  <Text
                    testID="cancelText"
                    style={[
                      styles.textSpinner,
                      { color: Colors.CancelButton },
                      this.props.cancelTextStyle,
                    ]}
                  >
                    {this.props.cancelText || 'Cancel'}
                  </Text>
                </TouchableOpacity>
              ) : null}
            </View>
          </View>
        </Modal>
      </View>
    );
  }
}

export default CrossBusyIndicator;
