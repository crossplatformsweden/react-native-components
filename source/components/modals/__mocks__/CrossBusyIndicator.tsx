export const CrossBusyIndicator = 'View';

export default CrossBusyIndicator;
