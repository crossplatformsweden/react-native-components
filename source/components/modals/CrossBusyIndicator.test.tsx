import React from 'react';
import TestRenderer, { ReactTestInstance } from 'react-test-renderer';
import _ from 'lodash';

import CrossBusyIndicator, { IBusyIndicatorProps } from './CrossBusyIndicator';
import { TouchableOpacity, Text } from 'react-native';
import Modal from 'react-native-modal';

jest.unmock('react-native');
jest.unmock('./CrossBusyIndicator');
jest.mock('react-native-modal', () => 'View');

function setup(options?: IBusyIndicatorProps | undefined) {
  const props: IBusyIndicatorProps = {
    isBusy: true,
    isCancelButtonVisible: true,
    type: 'BallIndicator',
    onCancel: () => console.log('Cancelled'),
    ...options,
  };

  const wrapper = TestRenderer.create(<CrossBusyIndicator {...props} />);

  return {
    props,
    wrapper,
  };
}

describe('components', () => {
  describe('<CrossBusyIndicator />', () => {
    it('Should render self and subcomponents', () => {
      const { wrapper } = setup();
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('Should not have <Modal /> when `isBusy` is `false`', () => {
      const { wrapper, props } = setup({
        type: 'DotIndicator',
        isBusy: false,
        isCancelButtonVisible: true,
      });
      expect(props.isBusy).toBeFalsy();

      const children = wrapper.root.findAllByProps({ visible: false });
      expect(children.length).toBe(0);
    });

    it('Should have <CrossSpinner /> of type "DotIndicator"', () => {
      const { wrapper } = setup({
        type: 'DotIndicator',
        isBusy: true,
        isCancelButtonVisible: true,
      });
      const child = wrapper.root.findByProps({
        type: 'DotIndicator',
      });
      expect(child).toBeDefined();
    });

    it('Should have <Text /> with value "Cancel"', () => {
      const { wrapper } = setup();
      const child = wrapper.root.findAllByType(Text).slice(-1)[0];
      expect(child.props.children).toBe('Cancel');
    });

    it('Cancel button onCancel should be called', () => {
      let called = false;
      const onCalled = () => (called = !called);

      const { wrapper } = setup({
        type: 'DotIndicator',
        isBusy: true,
        isCancelButtonVisible: true,
        onCancel: onCalled,
      });

      const child = wrapper.root.findByType(TouchableOpacity);
      child.props.onPress();
      expect(called).toBeTruthy();
    });

    it('Modal should have "swipeDirection" with value "up"', () => {
      const { wrapper } = setup({
        type: 'DotIndicator',
        isBusy: true,
        isCancelButtonVisible: true,
        modalProps: {
          swipeDirection: 'up',
        },
      });

      const child = wrapper.root.findByType(Modal)
        .children[0] as ReactTestInstance;
      expect(child.props.swipeDirection).toBe('up');
    });
  });
});
