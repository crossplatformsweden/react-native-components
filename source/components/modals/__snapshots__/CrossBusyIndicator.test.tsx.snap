// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <CrossBusyIndicator /> Should render self and subcomponents 1`] = `
<View
  testID="1"
>
  <View
    isVisible={true}
    onDismiss={[Function]}
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "justifyContent": "space-between",
          },
          Object {
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 20,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <View
          color="#057dc2"
          count={8}
          size={40}
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "center",
              "opacity": 1,
              "width": 50,
            }
          }
          testID="BallIndicator"
          type="BallIndicator"
        />
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "column",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#057dc2",
                "fontSize": 20,
                "fontWeight": "normal",
                "marginTop": 5,
                "opacity": 1,
                "textAlign": "center",
              },
              undefined,
            ]
          }
        >
          
        </Text>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#057dc2",
                  "fontSize": 20,
                  "fontWeight": "normal",
                  "marginTop": 5,
                  "opacity": 1,
                  "textAlign": "center",
                },
                Object {
                  "color": "#ff0000",
                },
                undefined,
              ]
            }
            testID="cancelText"
          >
            Cancel
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
